README.md
# fibx-router

## Installation

```javascript    
npm install @fibjs/fibx-router		
```

## Test				

```javascript
npm install --save-dev
npm test
```

## Example

```javascript
var app = require('@fibjs/fibx')();
var route = require('@fibjs/fibx-router')({
    fixSlash: true,
    simulation: true,
    size: 200
});

/** all 提供各个路由的统一执行体, 不提供路由能力 **/
route.all('/rube', function(next) {
    next('rube-dong');
});

route.get('/rube', function(p){
    this.body = p;
});

route.post('/rube', function(p){
    this.body = p;
});

route.get('/hello', function() {
    this.body = 'hello world';
});

route.get('/show/:name', function() {
    this.body = 'hello world ---' + this.params.name;
});

route.post('/show/:name', function() {
    this.body = 'hello world ---' + this.params.name;
});

app.use(function(next) {
    try {
        next();
    } catch (e) {
        this.body = 'error';
    }
});

app.use('(.*)', route.getAllRoute());

console.log('you can open browser, input 127.0.0.1:7759/hello    and more~');

app.listen(7759);
```

## Api

[API Doc](https://github.com/fibx/fibx-router/blob/master/doc/api.md)

## Next

* 对 route 使用 () 捕获参数的支持


CHANGELOG.md
* v1.4.0

```
1. 支持 fibjs 0.20.0 版本
3. 修复一些测试用例等兼容问题
```

test/debug.js
var coroutine = require("coroutine");
var http = require('http');
var app = require('@fibjs/fibx')();
var run = require('./server')(app);

coroutine.start(run);
coroutine.sleep(100);

var r = http.request('post', 'http://127.0.0.1:7758/slash/rube/');
console.log(r.read().toString());


test/index.js
var test = require('test');
test.setup();

var coroutine = require("coroutine");
var http = require('http');
var app = require('@fibjs/fibx')();
var run = require('./server')(app);
var isDebug = false;

app.use(function(next) {
    try {
        next();
    } catch (e) {
        if (isDebug){
            throw e;
        }
        this.body = 'error';
    }
});

coroutine.start(run);
coroutine.sleep(100);

describe('----------------------------fibx-router--------------------------\r\n', function() {

    describe('Method support', function() {
        it('get', function() {
            var r = http.request('get', 'http://127.0.0.1:7758/method');
            assert.equal(r.read().toString(), 'get');
        });

        it('post', function() {
            var r = http.request('post', 'http://127.0.0.1:7758/method');
            assert.equal(r.read().toString(), 'post');
        });

        it('delete', function() {
            var r = http.request('delete', 'http://127.0.0.1:7758/method');
            assert.equal(r.read().toString(), 'delete');
        });

        it('put', function() {
            var r = http.request('put', 'http://127.0.0.1:7758/method');
            assert.equal(r.read().toString(), 'put');
        });

        it('all', function() {
            var r = http.request('get', 'http://127.0.0.1:7758/aa');
            assert.equal(r.read().toString(), 'all2');
        });

        it('error method', function() {
            var r = http.request('rube', 'http://127.0.0.1:7758/aa');
            assert.equal(r.read().toString(), 'error');
        });

        it('put simulation', function() {
            var r = http.request('post', 'http://127.0.0.1:7758/sim/method', {body :"_method=put", headers: {"Content-type": "application/x-www-form-urlencoded"}});
            assert.equal(r.read().toString(), 'put');
        });

        it('delete simulation', function() {
            var r = http.request('post', 'http://127.0.0.1:7758/sim/method', {body: "_method=delete", headers: {"Content-type": "application/x-www-form-urlencoded"}});
            assert.equal(r.read().toString(), 'delete');
        });
    });

    describe('Router parse', function() {

        it('route base option', function() {
            var r = http.request('post', 'http://127.0.0.1:7758/base/papapa?h=hello_base');
            assert.equal(r.read().toString(), 'hello_base');
        });

        it('route fixSlash option', function() {
            var r = http.request('post', 'http://127.0.0.1:7758/slash/method/');
            assert.equal(r.read().toString(), 'slash ok');

            var r = http.request('post', 'http://127.0.0.1:7758/method/');
            assert.equal(r.read().toString(), 'error');

            var r = http.request('post', 'http://127.0.0.1:7758/slash/rube/');
            assert.equal(r.read().toString(), 'slash ok--rube');
        });

        it('error route', function() {
            var r = http.request('post', 'http://127.0.0.1:7758/base/papapsda?h=hello_base');
            assert.equal(r.read().toString(), 'error');

            var r = http.request('get', 'http://127.0.0.1:7758/metho');
            assert.equal(r.read().toString(), 'error');
        });

        it('route params', function() {
            var r = http.request('post', 'http://127.0.0.1:7758/params/rube');
            assert.equal(r.read().toString(), 'rube');

            var r = http.request('post', 'http://127.0.0.1:7758/params/rube/gg/dong');
            assert.equal(r.read().toString(), 'rube--dong');

            var r = http.request('post', 'http://127.0.0.1:7758/params/login/100888/check?name=dong');
            assert.equal(r.read().toString(), '100888--dong');

            var r = http.request('post', 'http://127.0.0.1:7758/params/login/gg/check?name=dong');
            assert.equal(r.read().toString(), 'login--check');
        });

        it('route params have base', function() {
            var r = http.request('post', 'http://127.0.0.1:7758/base/gogo/rube');
            assert.equal(r.read().toString(), 'rube');
        });

        it('multi route', function() {
            var r = http.request('post', 'http://127.0.0.1:7758/login/rubedong110');
            assert.equal(r.read().toString(), 'rubedong110--rubedong');

            var r = http.request('post', 'http://127.0.0.1:7758/sign/1688');
            assert.equal(r.read().toString(), 'regex1 ok');

            var r = http.request('post', 'http://127.0.0.1:7758/sign/rube/1688/dong');
            assert.equal(r.read().toString(), 'regex2 ok--rube--dong');

            var r = http.request('get', 'http://127.0.0.1:7758/sign/rube/hello/dong');
            assert.equal(r.read().toString(), 'regex3 ok--dong');

            var r = http.request('get', 'http://127.0.0.1:7758/sign/rube/dong/110/check');
            assert.equal(r.read().toString(), 'regex4 ok--dong');

            var r = http.request('post', 'http://127.0.0.1:7758/sign/f-u-c-k');
            assert.equal(r.read().toString(), 'error');
        });
    });

    describe('Multi executive body', function() {

        it('two executive body', function() {
            var r = http.request('get', 'http://127.0.0.1:7758/exec/two?h=hello');
            assert.equal(r.read().toString(), 'hello exec two');
        });

        it('more executive body', function() {
            var r = http.request('get', 'http://127.0.0.1:7758/exec/more?h=hello');
            assert.equal(r.read().toString(), 'hello exec more' + 5);
        });

        it('next can return value', function(){
            var r = http.request('get', 'http://127.0.0.1:7758/exec/next');
            assert.equal(r.read().toString(), 'next successful!');
        });

        it('executive body feature ok', function() {
            var r = http.request('get', 'http://127.0.0.1:7758/aa');
            assert.equal(r.read().toString(), 'all2');
        });
    });
});

test.run();
process.exit(0);


test/server.js
module.exports = function(app) {
    return function() {
        run(app);
    }
};

function run(app) {

    /** slash **/
    var routerSlash = require('../index')({
        fixSlash: true
    });
    routerSlash.post('/slash/method', function() {
        this.body = 'slash ok';
    });
    routerSlash.post('/slash/:method', function() {
        this.body = 'slash ok--' + this.params.method;
    });
    app.use('/slash(/.*)', routerSlash.getAllRoute());

    /** sim **/
    var routerSim = require('../index')({
        simulation: true
    });
    routerSim.put('/sim/method', function() {
        this.body = 'put';
    });
    routerSim.delete('/sim/method', function() {
        this.body = 'delete';
    });
    app.use('/sim(/.*)', routerSim.getAllRoute());

    /** base **/
    var routerBase = require('../index')({
        base: '/base'
    });
    routerBase.post('/papapa', function() {
        this.body = this.query.h;
    });
    routerBase.post('/gogo/:hello', function() {
        this.body = this.params.hello;
    });
    app.use('/base(/.*)', routerBase.getAllRoute());

    /** all **/
    var router = require('../index')();
    router.all('/aa', function(next) {
        this.state = {};
        this.state.number = 1;
        next();
    });
    router.get('/aa', function() {
        this.state.number++;
        this.body = 'all' + 2;
    });

    /** put, get, post, delete **/
    router.get('/method', function() {
        this.body = 'get';
    });
    router.post('/method', function() {
        this.body = 'post';
    });
    router.put('/method', function() {
        this.body = 'put';
    });
    router.delete('/method', function() {
        this.body = 'delete';
    });

    /** executive body **/
    router.get('/exec/two', function(next) {
        next(this.query.h);
    }, function(hello) {
        this.body = hello + ' exec two';
    });

    router.get('/exec/more', function(next) {
        next(1);
    }, function(n, next) {
        next(n + 1);
    }, function(n, next) {
        next(n + 1);
    }, function(n, next) {
        next(n + 1);
    }, function(n, next) {
        next(n + 1);
    }, function(n) {
        this.body = this.query.h + ' exec more' + n;
    });

    router.get('/exec/next', function(next) {
        this.body = next();
    }, function(){
        return 'next successful!'
    });

    /** params **/
    router.post('/params/:hello', function() {
        this.body = this.params.hello;
    });

    router.post('/params/:hello/gg/:good', function() {
        this.body = this.params.hello + '--' + this.params.good;
    });

    router.post('/params/login/:userID/check', function() {
        this.body = this.params.userID + '--' + this.query.name;
    });

    /** multi route **/
    router.post('/login/:userID', function(next) {
        next(this.params.userID + '--rubedong');
    }, function(params) {
        this.body = params;
    });
    router.post('/sign/[1-9]+', function() {
        this.body = 'regex1 ok';
    });
    router.post('/sign/:hello/[1-9]+/:good', function() {
        this.body = 'regex2 ok--' + this.params.hello + '--' + this.params.good;
    });
    router.get('/sign/[a-z]+/hello/:hello', function() {
        this.body = 'regex3 ok--' + this.params.hello;
    });
    router.get('/sign/[a-z]+/:hello/[0-9]+/check', function() {
        this.body = 'regex4 ok--' + this.params.hello;
    });

    app.use('(.*)', router.getAllRoute());

    app.listen(7758);
}

index.js
'use strict';

var util = require('util');

var error = require('./lib/error');
var connect = require('./lib/connect');
/**
 * @param option Object
 * - simulation: boolean    whether support put and delete simulation
 * - size: number           to save path cache size
 * - base: string           base path
 * - fixSlash: boolean      if true the request 127.0.0.1/hello/ will be ok
 * @constructor
 */
function Router(option) {

    this._option = option;
    this._cache = option && option.size ? new util.LruCache(option.size) : new util.LruCache(100);
    this._base = option && option.base ? option.base : '';
    this._pathParams = {};
    this._map = {
        get: {},
        post: {},
        delete: {},
        put: {},
        all: {}
    }
}

Router.prototype = {
    getAllRoute: function() {

        var that = this;

        return function(next) {

            var method = this.method.toLowerCase();
            //var methodSim = this.query._method || this.form._method;
            var methodSim = this.form._method;
            var path = this.path, regexPath;

            var l = path.length - 1;
            if (that._option
                && that._option.fixSlash
                && path.lastIndexOf('/') == l) {
                path = path.substr(0, l);
            }

            method = filterMethod.call(that, method, methodSim);
            regexPath = findPath.call(that, method, path);
            extractParams.call(this, that, regexPath, path);
            exec.call(this, that, method, regexPath);

            next && next();
        }
    }
};

var methodList = ['get', 'post', 'put', 'delete'];

methodList.concat(['all']).forEach(function(m) {

    Router.prototype[m] = function() {

        var argv = Array.prototype.slice.call(arguments);
        var regex = routeRegex.call(this, argv.shift());
        this._map[m][regex] = argv;
    }
});

module.exports = function(option) {
    return new Router(option);
};

function extractParams(that, regexPath, requestPath) {

    var _params = that._pathParams[regexPath];
    if (!_params || _params.length === 0) {
        return;
    }

    this.params = this.params || {};
    var regex = new RegExp('^' + that._base + regexPath + '$'),
        params = regex.exec(requestPath),
        _this = this;

    params.shift();
    params.forEach(function(p, idx) {
        _this.params[_params[idx]] = p;
    });
}

function exec(that, method, regexPath) {

    var executive = that._map[method][regexPath];
    if (that._map['all'][regexPath]) {
        executive = that._map['all'][regexPath].concat(executive);
    }
    connect.call(this, executive).call(this);
}

function filterMethod(method, methodSim) {

    if (!!!method) {
        throw error.P('not found method');
    }

    if (methodList.indexOf(method) != -1 && !!!methodSim) {
        return method;
    } else if (method === 'post' && (this._option && this._option.simulation)
        && (methodSim === 'put' || methodSim === 'delete')) {
        return methodSim;
    } else {
        throw error.P('not found method');
    }
}

function findPath(method, path) {

    var cacheResult = this._cache.get(path);
    if (cacheResult && cacheResult.m === method) {
        return cacheResult.p;
    } else {
        for (var _regex in this._map[method]) {
            var regex = new RegExp('^' + this._base + _regex + '$');                     //TODO: better
            if (regex.test(path)) {
                this._cache.set(path, {
                    m: method,
                    p: _regex
                });
                return _regex;
            }
        }
    }
    throw error.P('not found path');
}

function routeRegex(regex) {

    if (regex.indexOf('^') === 0) {
        regex = regex.substr(1);
    }
    if (regex.indexOf('$') === regex.length - 1) {
        regex = regex.substr(0, regex.length - 1);
    }

    var paramsRegex = /:[0-9a-zA-Z]+/g;
    var p;
    var params = [];
    while (p = paramsRegex.exec(regex)) {
        params.push(p[0].substr(1));
    }
    regex = regex.replace(/:[0-9a-zA-Z]+/g, '([^\/]+)');
    this._pathParams[regex] = params;

    return regex;
}


example/sample.js
var app = require('@fibjs/fibx')();
var route = require('../index')({
    fixSlash: true,
    simulation: true,
    size: 200
});

/** all 提供各个路由的统一执行体, 不提供路由能力 **/
route.all('/rube', function(next) {
    next('rube-dong');
});

route.get('/rube', function(p){
    this.body = p;
});

route.post('/rube', function(p){
    this.body = p;
});

route.get('/hello', function() {
    this.body = 'hello world';
});

route.get('/show/:name', function() {
    this.body = 'hello world ---' + this.params.name;
});

route.post('/show/:name', function() {
    this.body = 'hello world ---' + this.params.name;
});

app.use(function(next) {
    try {
        next();
    } catch (e) {
        this.body = 'error';
    }
});

app.use('/', route.getAllRoute());

console.log('you can open browser, input 127.0.0.1:7759/hello    and more~');

app.listen(7759);
